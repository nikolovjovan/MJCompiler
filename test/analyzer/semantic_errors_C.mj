// level C semantic errors
program level_c_semantic_errors
    // constant declarations
    const int a = 20, b = -1, c = +25;
    const char d = 'a';
    const bool e = true, f = false;
    const int A = 1, B = 2, C = 3;
    const char X = 'x', Y = 'y';
    const bool VALID = true;

    // variable declarations
    int var_a, var_b[];
    char var_c, var_d[], var_e;
    bool var_f;
    int a, b[], c;
    Point p1, p2;

    // class declarations
    class Point extends IPoint {
        public int x, y;
        {
            public int get_x() { return x; }
            public int diff_x(IPoint a) { return a.get_x() - x; }
        }
    }

    // abstract class declarations
    abstract class IPoint {
        private int d;
        {
            public abstract int get_x();
            public abstract int diff_x(IPoint a);
            private abstract void test(); // error: abstract method cannot be declared private
        }
    }

    const char SPACE = ' ';

    int x, y;

    abstract class Object {
        {
            public abstract void println();
        }
    }

    class Point extends Object
    {
        private int x, y;
        {
            public int getX() { return x; }
            public void setX(int x) { this.x = x; }

            public int getY() { return y; }
            public void setY(int y) { this.y = y; }

            public void println() {
                print(d); // error: superclass private field access forbidden
                print(x);
                print(SPACE);
                print(y);
                print(eol);
            }

            private void test() {} // error: duplicate name, private inherited method cannot be overridden
        }
    }

    abstract class abstract_class { // abstract inner class
        public int abstract_class_public_field;
        protected char abstract_class_protected_field;
        private bool abstract_class_private_field;
        {
            public abstract void public_method(int a, int b);
            protected abstract void protected_method(char a);
            private abstract void private_method(bool ind); // error: abstract method cannot be declared private
        }
    }

    class concrete_class extends abstract_class {
        // class variables
        public int class_var_a, class_var_b[];
        protected char class_var_c;
        private bool class_var_d;
        {
            // abstract method implementation
            public void public_method(int a, int b)
                int c;
            {
                abstract_class_public_field = 12; // superclass public field access
                abstract_class_protected_field = 'c'; // superclass protected field access
                abstract_class_private_field = false; // error: superclass private field access forbidden
                class_var_a = 15; // this class public field access
                class_var_c = 'd'; // this class protected field access
                class_var_d = false; // this class private field access
            }
            // abstract method implementation
            protected void protected_method(char a) {}
            // This is invalid because a method declared as private in any subclass cannot be overridden!
            private void private_method(bool ind) {}
        }
    }

    class test_class {
        // class variables
        public int class_var_a, class_var_b[];
        protected char class_var_c;
        private bool class_var_d;
        {
            public void test_method() {}
        }
    }
{
    int k()
        int c;
    {
        c = 1;
        return c + 2;
    }

    void printArray(int arr[])
        int it;
    {
        foreach (it : arr) {
            print(it);
            print(SPACE);
        }
        print(eol);
    }

    int calc(int a, int b, bool c) {
        if (c) return a + b;
        return a * b;
    }

    void test()
        // local variables
        int local_var_a, local_var_b[], i;
        char local_var_c;
        bool local_var_d, local_var_e[];
        test_class local_var_obj;
    {
        local_var_a = a; // int const usage
        local_var_b = new int[5]; // array instantiation
        local_var_c = d; // char const usage
        local_var_obj = new test_class;

        local_var_a = local_var_obj.class_var_a; // class field access
        local_var_b[3] = local_var_obj.class_var_b[5]; // class array field element access
        local_var_c = local_var_obj.class_var_c; // error: protected field access
        local_var_d = local_var_obj.class_var_d; // error: private field access
        local_var_obj.test_method(); // class method call

        local_var_a = calc(local_var_a, local_var_a, local_var_d); // test method call
        calc(); // invalid parameter count
        calc(local_var_a, local_var_d, local_var_a); // invalid argument type
        local_var_b = new int[c - a]; // const expr
        break; // break outside a loop
        for (i = 0; i < c - a; i++) {
            if (i % 2 == 0) local_var_c = true;
            else local_var_c = false;
            local_var_b[i] = calc(local_var_a, local_var_b[i], local_var_c); // invalid argument type
            continue;
        }
    }

    void main()
        int i, arr[];
        Point p;
        int a, b;
        char c;
        bool ind;
        abstract_class var;
    {
        a = 4;
        b = 15;
        c = 'b';
        ind = false;
        var = new concrete_class; // concrete class instantiation
        var.class_var_a = 5; // error: cannot access subclass public field from a superclass pointer
        var.abstract_class_public_field = -5; // superclass public field access
        var.class_var_c = 'b'; // error: cannot access subclass protected field access forbidden
        var.class_var_d = false; // error: cannot access subclass private field access forbidden
        var.public_method(a, b); // public method call
        var.protected_method(b); // error: protected method call forbidden
        var.private_method(c); // error: private method call forbidden
        y = 2;
        print(y); print(eol);
        x = y + 2;
        print(x); print(eol);
        y = k() + 3;
        print(y);
        p = new Point;
        p.x = x;
        p.y = y;
        p.printPoint();
        arr = new int[6];
        for (i = 0; i < len(arr); i++) {
            arr[i] = i * 7 - 2;
        }
        printArray(arr);
    }
}