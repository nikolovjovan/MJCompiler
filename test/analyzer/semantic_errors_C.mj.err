Semantic analysis completed with 29 error(s)!
