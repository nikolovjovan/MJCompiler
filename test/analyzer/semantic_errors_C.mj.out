Syntax analysis completed without errors.
Starting semantic analysis...
Found class 'Point' field usage 'x' at line 22! Symbol node: Field 'x': type = 'int', adr = 1, level = 1, parent = 'Point', access modifier = public
Found symbolic constant usage 'a' at line 23! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found class 'Point' field usage 'x' at line 23! Symbol node: Field 'x': type = 'int', adr = 1, level = 1, parent = 'Point', access modifier = public
Found abstract class usage 'IPoint' at line 32! Symbol node: Type 'IPoint': type = 'abstract class', adr = 0, level = 0
Found abstract class usage 'Object' at line 47! Symbol node: Type 'Object': type = 'abstract class', adr = 0, level = 0
Found class 'Point' field usage 'x' at line 51! Symbol node: Field 'x': type = 'int', adr = 1, level = 1, parent = 'Point', access modifier = private
Found local variable usage 'this' at line 52! Symbol node: Variable 'this': , adr = 0, level = 1
Found class 'Point' field usage 'x' at line 52! Symbol node: Field 'x': type = 'int', adr = 1, level = 1, parent = 'Point', access modifier = private
Found formal parameter usage 'x' at line 52! Symbol node: Formal parameter 'x': type = 'int', adr = 1, level = 1
Found class 'Point' field usage 'y' at line 54! Symbol node: Field 'y': type = 'int', adr = 2, level = 1, parent = 'Point', access modifier = private
Found local variable usage 'this' at line 55! Symbol node: Variable 'this': , adr = 0, level = 1
Found class 'Point' field usage 'y' at line 55! Symbol node: Field 'y': type = 'int', adr = 2, level = 1, parent = 'Point', access modifier = private
Found formal parameter usage 'y' at line 55! Symbol node: Formal parameter 'y': type = 'int', adr = 1, level = 1
Found symbolic constant usage 'd' at line 58! Symbol node: Constant 'd': type = 'char', adr = 97, level = 0
Found class 'Point' field usage 'x' at line 59! Symbol node: Field 'x': type = 'int', adr = 1, level = 1, parent = 'Point', access modifier = private
Found symbolic constant usage 'SPACE' at line 60! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found class 'Point' field usage 'y' at line 61! Symbol node: Field 'y': type = 'int', adr = 2, level = 1, parent = 'Point', access modifier = private
Found symbolic constant usage 'eol' at line 62! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found abstract class usage 'abstract_class' at line 80! Symbol node: Type 'abstract_class': type = 'abstract class', adr = 0, level = 0
Found class 'abstract_class' field usage 'abstract_class_public_field' at line 90! Symbol node: Field 'abstract_class_public_field': type = 'int', adr = 1, level = 1, parent = 'abstract_class', access modifier = public
Found class 'abstract_class' field usage 'abstract_class_protected_field' at line 91! Symbol node: Field 'abstract_class_protected_field': type = 'char', adr = 2, level = 1, parent = 'abstract_class', access modifier = protected
Found class 'concrete_class' field usage 'class_var_a' at line 93! Symbol node: Field 'class_var_a': type = 'int', adr = 4, level = 1, parent = 'concrete_class', access modifier = public
Found class 'concrete_class' field usage 'class_var_c' at line 94! Symbol node: Field 'class_var_c': type = 'char', adr = 6, level = 1, parent = 'concrete_class', access modifier = protected
Found class 'concrete_class' field usage 'class_var_d' at line 95! Symbol node: Field 'class_var_d': type = 'bool', adr = 7, level = 1, parent = 'concrete_class', access modifier = private
Found local variable usage 'c' at line 117! Symbol node: Variable 'c': type = 'int', adr = 0, level = 1
Found local variable usage 'c' at line 118! Symbol node: Variable 'c': type = 'int', adr = 0, level = 1
Found formal parameter usage 'arr' at line 124! Symbol node: Formal parameter 'arr': type = 'array of int', adr = 0, level = 1
Found local variable usage 'it' at line 125! Symbol node: Variable 'it': type = 'int', adr = 1, level = 1
Found symbolic constant usage 'SPACE' at line 126! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found symbolic constant usage 'eol' at line 128! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found formal parameter usage 'c' at line 132! Symbol node: Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found formal parameter usage 'a' at line 132! Symbol node: Formal parameter 'a': type = 'int', adr = 0, level = 1
Found formal parameter usage 'b' at line 132! Symbol node: Formal parameter 'b': type = 'int', adr = 1, level = 1
Found formal parameter usage 'a' at line 133! Symbol node: Formal parameter 'a': type = 'int', adr = 0, level = 1
Found formal parameter usage 'b' at line 133! Symbol node: Formal parameter 'b': type = 'int', adr = 1, level = 1
Found local variable usage 'local_var_a' at line 143! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found symbolic constant usage 'a' at line 143! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found local variable usage 'local_var_b' at line 144! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found local variable usage 'local_var_c' at line 145! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found symbolic constant usage 'd' at line 145! Symbol node: Constant 'd': type = 'char', adr = 97, level = 0
Found local variable usage 'local_var_obj' at line 146! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found local variable usage 'local_var_a' at line 148! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_obj' at line 148! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found class 'test_class' field usage 'class_var_a' at line 148! Symbol node: Field 'class_var_a': type = 'int', adr = 1, level = 1, parent = 'test_class', access modifier = public
Found local variable usage 'local_var_b' at line 149! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found array element access 'local_var_b[3]' at line 149! Symbol node: Array element 'local_var_b[3]': type = 'int', adr = -1, level = -1
Found local variable usage 'local_var_obj' at line 149! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found class 'test_class' field usage 'class_var_b' at line 149! Symbol node: Field 'class_var_b': type = 'array of int', adr = 2, level = 1, parent = 'test_class', access modifier = public
Found array element access 'local_var_obj.class_var_b[5]' at line 149! Symbol node: Array element 'local_var_obj.class_var_b[5]': type = 'int', adr = -1, level = -1
Found local variable usage 'local_var_c' at line 150! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found local variable usage 'local_var_obj' at line 150! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found local variable usage 'local_var_d' at line 151! Symbol node: Variable 'local_var_d': type = 'bool', adr = 4, level = 1
Found local variable usage 'local_var_obj' at line 151! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found local variable usage 'local_var_obj' at line 152! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found class 'test_class' field usage 'test_method' at line 152! Symbol node: Method 'test_method': type = 'void', adr = 0, level = 1, parent = 'test_class', access modifier = public
  Variable 'this': , adr = 0, level = 1
Found local variable usage 'local_var_a' at line 154! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found global method call 'calc' at line 154! Symbol node: Method 'calc': type = 'int', adr = 0, level = 3
  Formal parameter 'a': type = 'int', adr = 0, level = 1
  Formal parameter 'b': type = 'int', adr = 1, level = 1
  Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found local variable usage 'local_var_a' at line 154! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_a' at line 154! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_d' at line 154! Symbol node: Variable 'local_var_d': type = 'bool', adr = 4, level = 1
Found global method call 'calc' at line 155! Symbol node: Method 'calc': type = 'int', adr = 0, level = 3
  Formal parameter 'a': type = 'int', adr = 0, level = 1
  Formal parameter 'b': type = 'int', adr = 1, level = 1
  Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found global method call 'calc' at line 156! Symbol node: Method 'calc': type = 'int', adr = 0, level = 3
  Formal parameter 'a': type = 'int', adr = 0, level = 1
  Formal parameter 'b': type = 'int', adr = 1, level = 1
  Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found local variable usage 'local_var_a' at line 156! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_d' at line 156! Symbol node: Variable 'local_var_d': type = 'bool', adr = 4, level = 1
Found local variable usage 'local_var_a' at line 156! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_b' at line 157! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found symbolic constant usage 'c' at line 157! Symbol node: Constant 'c': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'a' at line 157! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found local variable usage 'i' at line 159! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found local variable usage 'i' at line 159! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found symbolic constant usage 'c' at line 159! Symbol node: Constant 'c': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'a' at line 159! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found local variable usage 'i' at line 159! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found for loop 'none' at line 159!
Found local variable usage 'i' at line 160! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found local variable usage 'local_var_c' at line 160! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found local variable usage 'local_var_c' at line 161! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found local variable usage 'local_var_b' at line 162! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found local variable usage 'i' at line 162! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found array element access 'i[i]' at line 162! Symbol node: Array element 'i[i]': type = 'int', adr = -1, level = -1
Found global method call 'calc' at line 162! Symbol node: Method 'calc': type = 'int', adr = 0, level = 3
  Formal parameter 'a': type = 'int', adr = 0, level = 1
  Formal parameter 'b': type = 'int', adr = 1, level = 1
  Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found local variable usage 'local_var_a' at line 162! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_b' at line 162! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found local variable usage 'i' at line 162! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found array element access 'i[i]' at line 162! Symbol node: Array element 'i[i]': type = 'int', adr = -1, level = -1
Found local variable usage 'local_var_c' at line 162! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found abstract class usage 'abstract_class' at line 173! Symbol node: Type 'abstract_class': type = 'abstract class', adr = 0, level = 0
Found local variable usage 'a' at line 175! Symbol node: Variable 'a': type = 'int', adr = 3, level = 1
Found local variable usage 'b' at line 176! Symbol node: Variable 'b': type = 'int', adr = 4, level = 1
Found local variable usage 'c' at line 177! Symbol node: Variable 'c': type = 'char', adr = 5, level = 1
Found local variable usage 'ind' at line 178! Symbol node: Variable 'ind': type = 'bool', adr = 6, level = 1
Found local variable usage 'var' at line 179! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found local variable usage 'var' at line 180! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found local variable usage 'var' at line 181! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found class 'abstract_class' field usage 'abstract_class_public_field' at line 181! Symbol node: Field 'abstract_class_public_field': type = 'int', adr = 1, level = 1, parent = 'abstract_class', access modifier = public
Found local variable usage 'var' at line 182! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found local variable usage 'var' at line 183! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found local variable usage 'var' at line 184! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found class 'abstract_class' field usage 'public_method' at line 184! Symbol node: Abstract method 'public_method': type = 'void', adr = 0, level = 3, parent = 'abstract_class', access modifier = public
  Variable 'this': , adr = 0, level = 1
  Formal parameter 'a': type = 'int', adr = 1, level = 1
  Formal parameter 'b': type = 'int', adr = 2, level = 1
Found local variable usage 'a' at line 184! Symbol node: Variable 'a': type = 'int', adr = 3, level = 1
Found local variable usage 'b' at line 184! Symbol node: Variable 'b': type = 'int', adr = 4, level = 1
Found local variable usage 'var' at line 185! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found local variable usage 'b' at line 185! Symbol node: Variable 'b': type = 'int', adr = 4, level = 1
Found local variable usage 'var' at line 186! Symbol node: Variable 'var': type = 'class', adr = 7, level = 1
Found local variable usage 'c' at line 186! Symbol node: Variable 'c': type = 'char', adr = 5, level = 1
Found global variable usage 'y' at line 187! Symbol node: Variable 'y': type = 'int', adr = 7, level = 0
Found global variable usage 'y' at line 188! Symbol node: Variable 'y': type = 'int', adr = 7, level = 0
Found symbolic constant usage 'eol' at line 188! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'x' at line 189! Symbol node: Variable 'x': type = 'int', adr = 6, level = 0
Found global variable usage 'y' at line 189! Symbol node: Variable 'y': type = 'int', adr = 7, level = 0
Found global variable usage 'x' at line 190! Symbol node: Variable 'x': type = 'int', adr = 6, level = 0
Found symbolic constant usage 'eol' at line 190! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'y' at line 191! Symbol node: Variable 'y': type = 'int', adr = 7, level = 0
Found global method call 'k' at line 191! Symbol node: Method 'k': type = 'int', adr = 0, level = 0
  Variable 'c': type = 'int', adr = 0, level = 1
Found global variable usage 'y' at line 192! Symbol node: Variable 'y': type = 'int', adr = 7, level = 0
Found local variable usage 'p' at line 193! Symbol node: Variable 'p': type = 'class', adr = 2, level = 1
Found local variable usage 'p' at line 194! Symbol node: Variable 'p': type = 'class', adr = 2, level = 1
Found class 'Point' field usage 'x' at line 194! Symbol node: Field 'x': type = 'int', adr = 1, level = 1, parent = 'Point', access modifier = public
Found global variable usage 'x' at line 194! Symbol node: Variable 'x': type = 'int', adr = 6, level = 0
Found local variable usage 'p' at line 195! Symbol node: Variable 'p': type = 'class', adr = 2, level = 1
Found class 'Point' field usage 'y' at line 195! Symbol node: Field 'y': type = 'int', adr = 2, level = 1, parent = 'Point', access modifier = public
Found global variable usage 'y' at line 195! Symbol node: Variable 'y': type = 'int', adr = 7, level = 0
Found local variable usage 'p' at line 196! Symbol node: Variable 'p': type = 'class', adr = 2, level = 1
Found local variable usage 'arr' at line 197! Symbol node: Variable 'arr': type = 'array of int', adr = 1, level = 1
Found local variable usage 'i' at line 198! Symbol node: Variable 'i': type = 'int', adr = 0, level = 1
Found local variable usage 'i' at line 198! Symbol node: Variable 'i': type = 'int', adr = 0, level = 1
Found global method call 'len' at line 198! Symbol node: Method 'len': type = 'int', adr = 0, level = 1
  Formal parameter 'arr': type = 'array of none', adr = 0, level = 1
Found local variable usage 'arr' at line 198! Symbol node: Variable 'arr': type = 'array of int', adr = 1, level = 1
Found local variable usage 'i' at line 198! Symbol node: Variable 'i': type = 'int', adr = 0, level = 1
Found for loop 'none' at line 198!
Found local variable usage 'arr' at line 199! Symbol node: Variable 'arr': type = 'array of int', adr = 1, level = 1
Found local variable usage 'i' at line 199! Symbol node: Variable 'i': type = 'int', adr = 0, level = 1
Found array element access 'i[i]' at line 199! Symbol node: Array element 'i[i]': type = 'int', adr = -1, level = -1
Found local variable usage 'i' at line 199! Symbol node: Variable 'i': type = 'int', adr = 0, level = 1
Found global method call 'printArray' at line 201! Symbol node: Method 'printArray': type = 'void', adr = 0, level = 1
  Formal parameter 'arr': type = 'array of int', adr = 0, level = 1
  Variable 'it': type = 'int', adr = 1, level = 1
Found local variable usage 'arr' at line 201! Symbol node: Variable 'arr': type = 'array of int', adr = 1, level = 1
======================================== SYMBOL TABLE DUMP =========================================
Type 'int': type = 'int', adr = -1, level = -1
Type 'char': type = 'char', adr = -1, level = -1
Type 'bool': type = 'bool', adr = -1, level = -1
Constant 'null': type = 'class', adr = 0, level = 0
Constant 'eol': type = 'char', adr = 10, level = 0
Method 'chr': type = 'char', adr = 0, level = 1
  Formal parameter 'i': type = 'int', adr = 0, level = 1
Method 'ord': type = 'int', adr = 0, level = 1
  Formal parameter 'ch': type = 'char', adr = 0, level = 1
Method 'len': type = 'int', adr = 0, level = 1
  Formal parameter 'arr': type = 'array of none', adr = 0, level = 1
Program 'level_c_semantic_errors': type = 'none', adr = 0, level = 1
  Constant 'a': type = 'int', adr = 20, level = 0
  Constant 'b': type = 'int', adr = -1, level = 0
  Constant 'c': type = 'int', adr = 25, level = 0
  Constant 'd': type = 'char', adr = 97, level = 0
  Constant 'e': type = 'bool', adr = 1, level = 0
  Constant 'f': type = 'bool', adr = 0, level = 0
  Constant 'A': type = 'int', adr = 1, level = 0
  Constant 'B': type = 'int', adr = 2, level = 0
  Constant 'C': type = 'int', adr = 3, level = 0
  Constant 'X': type = 'char', adr = 120, level = 0
  Constant 'Y': type = 'char', adr = 121, level = 0
  Constant 'VALID': type = 'bool', adr = 1, level = 0
  Variable 'var_a': type = 'int', adr = 0, level = 0
  Variable 'var_b': type = 'array of int', adr = 1, level = 0
  Variable 'var_c': type = 'char', adr = 2, level = 0
  Variable 'var_d': type = 'array of char', adr = 3, level = 0
  Variable 'var_e': type = 'char', adr = 4, level = 0
  Variable 'var_f': type = 'bool', adr = 5, level = 0
  Type 'Point': type = 'class', adr = 0, level = 0
  Type 'IPoint': type = 'abstract class', adr = 0, level = 0
  Constant 'SPACE': type = 'char', adr = 32, level = 0
  Variable 'x': type = 'int', adr = 6, level = 0
  Variable 'y': type = 'int', adr = 7, level = 0
  Type 'Object': type = 'abstract class', adr = 0, level = 0
  Type 'abstract_class': type = 'abstract class', adr = 0, level = 0
  Type 'concrete_class': type = 'class', adr = 0, level = 0
  Type 'test_class': type = 'class', adr = 0, level = 0
  Method 'k': type = 'int', adr = 0, level = 0
    Variable 'c': type = 'int', adr = 0, level = 1
  Method 'printArray': type = 'void', adr = 0, level = 1
    Formal parameter 'arr': type = 'array of int', adr = 0, level = 1
    Variable 'it': type = 'int', adr = 1, level = 1
  Method 'calc': type = 'int', adr = 0, level = 3
    Formal parameter 'a': type = 'int', adr = 0, level = 1
    Formal parameter 'b': type = 'int', adr = 1, level = 1
    Formal parameter 'c': type = 'bool', adr = 2, level = 1
  Method 'test': type = 'void', adr = 0, level = 0
    Variable 'local_var_a': type = 'int', adr = 0, level = 1
    Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
    Variable 'i': type = 'int', adr = 2, level = 1
    Variable 'local_var_c': type = 'char', adr = 3, level = 1
    Variable 'local_var_d': type = 'bool', adr = 4, level = 1
    Variable 'local_var_e': type = 'array of bool', adr = 5, level = 1
    Variable 'local_var_obj': type = 'class', adr = 6, level = 1
  Method 'main': type = 'void', adr = 0, level = 0
    Variable 'i': type = 'int', adr = 0, level = 1
    Variable 'arr': type = 'array of int', adr = 1, level = 1
    Variable 'p': type = 'class', adr = 2, level = 1
    Variable 'a': type = 'int', adr = 3, level = 1
    Variable 'b': type = 'int', adr = 4, level = 1
    Variable 'c': type = 'char', adr = 5, level = 1
    Variable 'ind': type = 'bool', adr = 6, level = 1
    Variable 'var': type = 'class', adr = 7, level = 1

